File: src/main/scala/planstack/anml/model/concrete/TemporalConstraint.scala
package planstack.anml.model.concrete

import planstack.anml.model.{AnmlProblem, LActRef, Context}
import planstack.anml.model.abs.{AbstractTemporalConstraint}
import planstack.anml.model.concrete.time.TimepointRef




case class TemporalConstraint(tp1:TPRef, op:String, tp2:TPRef, plus:Integer) {
  require(op == "<" || op == "=")

  override def toString = "%s %s %s + %s".format(tp1, op, tp2, plus)
}

object TemporalConstraint {

  def apply(pb:AnmlProblem, context:Context, abs:AbstractTemporalConstraint) = {
    new TemporalConstraint(TimepointRef(pb, context, abs.tp1), abs.op, TimepointRef(pb, context, abs.tp2), abs.plus)
  }
}

File: planning/src/main/java/fape/core/planning/search/flaws/resolvers/TemporalConstraint.java
package fape.core.planning.search.flaws.resolvers;

import fape.core.planning.planner.APlanner;
import fape.core.planning.states.State;
import planstack.anml.model.concrete.TPRef;

import java.util.Collections;
import java.util.List;

/**
 * Enforces a temporal constraints between two time points.
 */
public class TemporalConstraint extends Resolver {

    public final List<TPRef> firsts, seconds;
    public final int min, max;

    public TemporalConstraint(List<TPRef> firsts, List<TPRef> seconds, int min, int max) {
        this.firsts = firsts;
        this.seconds = seconds;
        this.min = min;
        this.max = max;
    }

    public TemporalConstraint(TPRef first, TPRef second, int min, int max) {
        this.firsts = Collections.singletonList(first);
        this.seconds = Collections.singletonList(second);
        this.min = min;
        this.max = max;
    }

    @Override
    public boolean apply(State st, APlanner planner, boolean isFastForwarding) {
        for(TPRef first : firsts)
            for(TPRef second : seconds)
                st.enforceConstraint(first, second, min, max);
        return true;
    }

    @Override
    public int compareWithSameClass(Resolver e) {
        assert e instanceof TemporalConstraint;
        TemporalConstraint o = (TemporalConstraint) e;
        for(TPRef first : firsts)
            for(TPRef ofirst : o.firsts)
                if(first != ofirst)
                    return first.id() - ofirst.id();
        for(TPRef second : seconds)
            for(TPRef osecond : o.seconds)
                if(second != osecond)
                    return second.id() - osecond.id();
        if(min != o.min)
            return min - o.min;

        assert max != o.max : "Error: comparing two identical resolvers.";
        return max - o.max;
    }
}
